File: src/2018/05.ts
// https://adventofcode.com/2018/day/5

import { read } from "../lib/read";

function _getPolymer(): string {
  return read("05");
}

function _match(char1: string, char2: string): boolean {
  return char1 !== char2 && char1.toLowerCase() === char2.toLowerCase();
}

function _reactedLength(polymer: string): number {
  let reactionCount = 0;
  let unmatchedIndexes = [];
  let index1;
  let index2;
  let lastPassWasMatch = false;
  let run = true;

  while (run) {
    // initial pass
    if (index1 === undefined && index2 === undefined) {
      index1 = 0;
      index2 = 1;
    } else {
      if (lastPassWasMatch) {
        index1 = unmatchedIndexes[unmatchedIndexes.length - 1] || index2;
        index2++;
      } else {
        index1 = index2;
        index2++;
      }
    }

    if (index2 >= polymer.length) {
      run = false;
      break;
    }

    const char1 = polymer[index1];
    const char2 = polymer[index2];

    if (_match(char1, char2)) {
      reactionCount++;

      if (lastPassWasMatch) {
        unmatchedIndexes.pop();
      }

      lastPassWasMatch = true;
    } else {
      lastPassWasMatch = false;
      if (unmatchedIndexes.indexOf(index1) === -1) {
        unmatchedIndexes.push(index1);
      }
    }
  }

  return polymer.length - reactionCount * 2;
}

function part1(): number {
  const polymer = _getPolymer();

  return _reactedLength(polymer);
}

function part2(): number {
  const polymer = _getPolymer();
  let shortestPolymerLength;

  for (let index = 65; index <= 90; index++) {
    const characterToRemove = String.fromCharCode(index);
    const shorterPolymer = polymer.replace(new RegExp(characterToRemove, "ig"), "");
    const reactedLength = _reactedLength(shorterPolymer);

    if (!shortestPolymerLength || reactedLength < shortestPolymerLength) {
      shortestPolymerLength = reactedLength;
    }
  }

  return shortestPolymerLength;
}

describe("05", () => {
  test("part 1", () => {
    expect(part1()).toBe(9526);
  });

  test("part 2", () => {
    expect(part2()).toBe(6694);
  });
});

File: rust/2018/src/bin/05.rs
// https://adventofcode.com/2018/day/5

fn reacted_length(polymer: &str, ignore: Option<u8>) -> usize {
    let mut reacted_polymer: Vec<u8> = vec![];

    for char in polymer.bytes() {
        if let Some(ignored_char) = ignore {
            if char == ignored_char || char + 32 == ignored_char {
                continue;
            }
        }

        if let Some(last_char) = reacted_polymer.last() {
            if char - 32 == *last_char || *last_char - 32 == char {
                reacted_polymer.pop();
            } else {
                reacted_polymer.push(char);
            }
        } else {
            reacted_polymer.push(char);
        }
    }

    reacted_polymer.len()
}

fn part1(input: &str) -> usize {
    reacted_length(input, None)
}

fn part2(input: &str) -> usize {
    (97u8..=122)
        .map(|char| reacted_length(input, Some(char)))
        .min()
        .unwrap()
}

fn main() {
    let input = include_str!("../../../../_input/2018/05.txt").trim_end();

    println!("part 1: {}", part1(input));
    println!("part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    const INPUT: &str = include_str!("../../../../_input/2018/05.txt");

    #[test]
    fn part1() {
        assert_eq!(super::part1(INPUT.trim_end()), 9526);
    }

    #[test]
    fn part2() {
        assert_eq!(super::part2(INPUT.trim_end()), 6694);
    }
}
